src/librustc/mir/visit.rs

```diff
@@ -880,125 +945,156 @@ pub enum TyContext {
+            }) |
+            PlaceContext::MaybeMutatingUse(MaybeMutatingUseContext::Projection(Mutability::Mut)) |
+            _ => false,
         }
     }
 
-    /// Returns true if this place context represents a use that does not change the value.
+    /// Returns `true` if this place context represents a use that does not change the value.
     pub fn is_nonmutating_use(&self) -> bool {
         match *self {
-            PlaceContext::Inspect |
-            PlaceContext::Borrow { kind: BorrowKind::Shared, .. } |
-            PlaceContext::Borrow { kind: BorrowKind::Shallow, .. } |
-            PlaceContext::Borrow { kind: BorrowKind::Unique, .. } |
-            PlaceContext::Projection(Mutability::Not) |
-            PlaceContext::Copy | PlaceContext::Move => true,
-
-            PlaceContext::Borrow { kind: BorrowKind::Mut { .. }, .. } | PlaceContext::Store |
-            PlaceContext::AsmOutput |
-            PlaceContext::Call | PlaceContext::Projection(Mutability::Mut) |
-            PlaceContext::Drop | PlaceContext::StorageLive | PlaceContext::StorageDead |
-            PlaceContext::Validate => false,
+            PlaceContext::NonMutatingUse(..) |
+            PlaceContext::MaybeMutatingUse(MaybeMutatingUseContext::Projection(Mutability::Not)) =>
+                true,
+            PlaceContext::MaybeMutatingUse(MaybeMutatingUseContext::Borrow { kind, .. }) =>
+                match kind {
+                    BorrowKind::Shared | BorrowKind::Shallow | BorrowKind::Unique => true,
+                    _ => false,
+                },
+            _ => false,
         }
     }
 
+    /// Returns `true` if this place context represents a use.
     pub fn is_use(&self) -> bool {
-        self.is_mutating_use() || self.is_nonmutating_use()
+        match *self {
+            PlaceContext::NonUse(..) => false,
+            _ => true,
+        }
+    }
+
+    /// Returns `true` if this place context represents an assignment statement.
+    pub fn is_place_assignment(&self) -> bool {
+        match *self {
+            PlaceContext::MutatingUse(MutatingUseContext::Store) |
+            PlaceContext::MutatingUse(MutatingUseContext::Call) |
+            PlaceContext::MutatingUse(MutatingUseContext::AsmOutput) => true,
+            PlaceContext::MutatingUse(MutatingUseContext::Drop) |
+            PlaceContext::NonMutatingUse(NonMutatingUseContext::Inspect) |
+            PlaceContext::MaybeMutatingUse(MaybeMutatingUseContext::Borrow { .. }) |
+            PlaceContext::MaybeMutatingUse(MaybeMutatingUseContext::Projection(..)) |
+            PlaceContext::NonMutatingUse(NonMutatingUseContext::Copy) |
+            PlaceContext::NonMutatingUse(NonMutatingUseContext::Move) |
+            PlaceContext::NonUse(NonUseContext::StorageLive) |
+            PlaceContext::NonUse(NonUseContext::StorageDead) |
+            PlaceContext::NonUse(NonUseContext::AscribeUserTy) |
+            PlaceContext::NonUse(NonUseContext::Validate) => false,
+        }
```

Comment:
What is this used for? It's a new classification, isn't it?

Reply:
I moved this function from another pass that I stumbled upon while making changes, it was used in a handful of places and was being imported from that pass.